README
NAME
    Devel::StackTrace - An object representing a stack trace

VERSION
    version 1.33

SYNOPSIS
      use Devel::StackTrace;

      my $trace = Devel::StackTrace->new();

      print $trace->as_string(); # like carp

      # from top (most recent) of stack to bottom.
      while ( my $frame = $trace->next_frame() ) {
          print "Has args\n" if $frame->hasargs();
      }

      # from bottom (least recent) of stack to top.
      while ( my $frame = $trace->prev_frame() ) {
          print "Sub: ", $frame->subroutine(), "\n";
      }

DESCRIPTION
    The "Devel::StackTrace" module contains two classes,
    C,Devel::StackTrace> and Devel::StackTrace::Frame. These objects
    encapsulate the information that can retrieved via Perl's "caller()"
    function, as well as providing a simple interface to this data.

    The "Devel::StackTrace" object contains a set of
    "Devel::StackTrace::Frame" objects, one for each level of the stack. The
    frames contain all the data available from "caller()".

    This code was created to support my Exception::Class::Base class (part
    of Exception::Class) but may be useful in other contexts.

'TOP' AND 'BOTTOM' OF THE STACK
    When describing the methods of the trace object, I use the words 'top'
    and 'bottom'. In this context, the 'top' frame on the stack is the most
    recent frame and the 'bottom' is the least recent.

    Here's an example:

      foo();  # bottom frame is here

      sub foo {
         bar();
      }

      sub bar {
         Devel::StackTrace->new();  # top frame is here.
      }

METHODS
    This class provide the following methods:

  Devel::StackTrace->new(%named_params)
    Returns a new Devel::StackTrace object.

    Takes the following parameters:

    *   frame_filter => $sub

        By default, Devel::StackTrace will include all stack frames before
        the call to its constructor.

        However, you may want to filter out some frames with more
        granularity than 'ignore_package' or 'ignore_class' allow.

        You can provide a subroutine which is called with the raw frame data
        for each frame. This is a hash reference with two keys, "caller",
        and "args", both of which are array references. The "caller" key is
        the raw data as returned by Perl's "caller()" function, and the
        "args" key are the subroutine arguments found in @DB::args.

        The filter should return true if the frame should be included, or
        false if it should be skipped.

    *   filter_frames_early => $boolean

        If this parameter is true, "frame_filter" will be called as soon as
        the stacktrace is created, and before refs are stringified (if
        "no_refs" is true), rather than being filtered lazily when
        Devel::StackTrace::Frame objects are first needed.

        This is useful if you want to filter based on the frame's arguments
        and want to be able to examine object properties, for example.

    *   ignore_package => $package_name OR \@package_names

        Any frames where the package is one of these packages will not be on
        the stack.

    *   ignore_class => $package_name OR \@package_names

        Any frames where the package is a subclass of one of these packages
        (or is the same package) will not be on the stack.

        Devel::StackTrace internally adds itself to the 'ignore_package'
        parameter, meaning that the Devel::StackTrace package is ALWAYS
        ignored. However, if you create a subclass of Devel::StackTrace it
        will not be ignored.

    *   skip_frames => $integer

        This will cause this number of stack frames to be excluded from top
        of the stack trace. This prevents the frames from being captured at
        all, and applies before the "frame_filter", "ignore_package", or
        "ignore_class" options, even with "filter_frames_early".

    *   no_refs => $boolean

        If this parameter is true, then Devel::StackTrace will not store
        references internally when generating stacktrace frames. This lets
        your objects go out of scope.

        Devel::StackTrace replaces any references with their stringified
        representation.

    *   no_args => $boolean

        If this parameter is true, then Devel::StackTrace will not store
        caller arguments in stack trace frames at all.

    *   respect_overload => $boolean

        By default, Devel::StackTrace will call "overload::AddrRef()" to get
        the underlying string representation of an object, instead of
        respecting the object's stringification overloading. If you would
        prefer to see the overloaded representation of objects in stack
        traces, then set this parameter to true.

    *   max_arg_length => $integer

        By default, Devel::StackTrace will display the entire argument for
        each subroutine call. Setting this parameter causes truncates each
        subroutine argument's string representation if it is longer than
        this number of characters.

    *   message => $string

        By default, Devel::StackTrace will use 'Trace begun' as the message
        for the first stack frame when you call "as_string". You can supply
        an alternative message using this option.

    *   indent => $boolean

        If this parameter is true, each stack frame after the first will
        start with a tab character, just like "Carp::confess()".

  $trace->next_frame()
    Returns the next Devel::StackTrace::Frame object on the stack, going
    down. If this method hasn't been called before it returns the first
    frame. It returns "undef" when it reaches the bottom of the stack and
    then resets its pointer so the next call to "$trace->next_frame()" or
    "$trace->prev_frame()" will work properly.

  $trace->prev_frame()
    Returns the next Devel::StackTrace::Frame object on the stack, going up.
    If this method hasn't been called before it returns the last frame. It
    returns undef when it reaches the top of the stack and then resets its
    pointer so the next call to "$trace->next_frame()" or
    "$trace->prev_frame()" will work properly.

  $trace->reset_pointer
    Resets the pointer so that the next call to "$trace->next_frame()" or
    "$trace->prev_frame()" will start at the top or bottom of the stack, as
    appropriate.

  $trace->frames()
    When this method is called with no arguments, it returns a list of
    Devel::StackTrace::Frame objects. They are returned in order from top
    (most recent) to bottom.

    This method can also be used to set the object's frames if you pass it a
    list of Devel::StackTrace::Frame objects.

    This is useful if you want to filter the list of frames in ways that are
    more complex than can be handled by the "$trace->filter_frames()"
    method:

      $stacktrace->frames( my_filter( $stacktrace->frames() ) );

  $trace->frame($index)
    Given an index, this method returns the relevant frame, or undef if
    there is no frame at that index. The index is exactly like a Perl array.
    The first frame is 0 and negative indexes are allowed.

  $trace->frame_count()
    Returns the number of frames in the trace object.

  $trace->as_string(\%p)
    Calls "$frame->as_string()" on each frame from top to bottom, producing
    output quite similar to the Carp module's cluck/confess methods.

    The optional "\%p" parameter only has one option. The "max_arg_length"
    parameter truncates each subroutine argument's string representation if
    it is longer than this number of characters.

SUPPORT
    Please submit bugs to the CPAN RT system at
    http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Devel%3A%3AStackTrace or
    via email at bug-devel-stacktrace@rt.cpan.org.

AUTHOR
    Dave Rolsky <autarch@urth.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2014 by Dave Rolsky.

    This is free software, licensed under:

      The Artistic License 2.0 (GPL Compatible)



Changes
2.00   2014-11-01

[BACKWARDS INCOMPATIBILITIES]

- The no_refs constructor parameter is now deprecated, and has been replace by
  a new unsafe_ref_capture parameter that defaults to false, meaning no
  references are captured by default. Capturing references by default caused
  too many issues that couldn't be worked around, including running DESTROY
  blocks multiple times on captured objects in the worst case.

- Removed support for the long-deprecated no_object_refs constructor parameter
  (deprecated in 2002!).


1.34   2014-06-26

- Fixed use of // operator (my use, not Graham's) in previous release.


1.33   2014-06-26

- Added a skip_frames option. This causes the stack trace to skip an arbitrary
  number of frames. Patch by Graham Knopp. PR #5.


1.32   2014-05-05

- Added a filter_frames_early option to filter frames before arguments are
  stringified. Added by Dagfinn Ilmari Mannsåker. PR #4.


1.31   2014-01-16

- No code changes, just doc updates, including documenting the as_string()
  method in Devel::StackTrace::Frame. Requested by Skef. RT #91575.


1.30   2012-11-19

- There was an eval which did not first localize $@ and $SIG{__DIE__}. This
  broke Plack::Middleware::StackTrace (and possibly other tihngs).


1.29   2012-11-16

- The Devel::StackTrace->frames() method is now read-write. This allows you to
  do more complex filtering of frames than is easily possible with the
  frame_filter argument to the constructor. Patch by David Cantrell.


1.28   2012-11-16

- Allow arguments to a trace's as_string method, specifically max_arg_length
  Patch by Ricardo Signes.

- Added a no_args option to the constructor in 1.26 but forgot to mention it
  in Changes. Requested by Scott J. Miller. RT #71482.


1.27  2011-01-16

- Skip some tests on 5.13.8+ that are no longer relevant because of a change
  in the Perl core. Reported by Andreas Koenig. RT #64828.


1.26  2010-10-15

- The as_string method did not localize $@ and $SIG{__DIE__} before doing an
  eval. Reported and tested by Marc Mims. RT #61072.


1.25  2010-09-06

- Devel::StackTraceFrame was not actually subclassing
  Devel::StackTrace::Frame. Patch by Tatsuhiko Miyagawa.


1.24  2010-09-03

- Version 1.23 was missing a $VERSION assignment. Reported by Sergei
  Vyshenski.

- Moved the frame object to its own file, and renamed it
  Devel::StackTrace::Frame. The old package name, Devel::StackTraceFrame, is
  now a subclass of the new package, to provide a backwards compatibility
  shim.


1.23  2010-08-27

- Added message and indent constructor parameters. Based on a patch by James
  Laver. RT #59830.


1.22  2009-07-15

- Apparently, overload::StrVal on older Perls (5.8.5, but not 5.8.8)
  tried to call a stringification method if it existed. So now,
  Devel::StackTrace just uses overload::AddrRef instead, which should
  always be safe. Reported by Michael Stevens. Fixes RT #47900.


1.21  2009-07-01

- Overloaded objects which didn't provide a stringification method
  cause Devel::StackTrace to die when respect_overload was
  true. Reported by Laurent Dami. RT #39533.

- Added a frame_filter option which allows for fine-grained control
  over what frames are included in a trace. Based on (but expanded)
  from a patch proposed by Florian Ragwitz. RT #47415.


1.20  2008-10-25

- The change in 1.15 to object creation broke the no_refs feature,
  causing references to be stored until the trace's frame objects were
  created.

* Exception::Class objects are always stringified by calling
  overload::StrVal().


1.1902  2008-07-16

- This release just contains another test fix.

- The new tests for bad utf-8 apparently fail with any Perl before
  5.8.8. Reported by Lee Heagney. RT #37702.


1.1901  2008-06-13

- This release just contains a test fix.

- The new tests for bad utf-8 fail with Perl 5.8.x where x <=
  6. Apparently, utf-8 was just more broken back then. Reported by
  Andreas Koenig's smokebots.


1.19  2008-06-13

- Dropped support for Perl 5.005.

- If a function was in stack trace had been called with invalid utf-8
  bytes, this could cause stringifying a stack trace to blow up when
  it tried to stringify that argument. We now catch those (and other)
  errors and simply put "(bad utf-8)" or "?" in the stringified
  argument list. Reported by Alex Vandiver.


1.18  2008-03-31

- Fix a test failure on Win32. No changes to the non-test code.


1.17  2008-03-30

- Added a max_arg_length parameter, which if set causes
  Devel::StackTrace to truncate long strings when printing out a
  frame. RT #33519. Patch by Ian Burrell. 


1.16  2008-02-02

- A test fix for bleadperl. The value of wantarray from caller() needs
  to be treated as a boolean, as opposed to expecting 0 (vs
  undef). RT #32583. Patch by Jerry Hedden.


1.15  2007-04-28

- Changed how objects are created in order to greatly speed up the
  constructor. Instead of processing all the stack trace data when the
  object is first created, this is delayed until it is needed. This
  was done in order to help speed up Exception::Class. There are cases
  where code may be throwing many exceptions but never examining the
  stack traces.

  Here is a representative benchmark of object construction for the
  old code versus the new code:

          Rate  old  new
    old 1764/s   -- -76%
    new 7353/s 317%   --


1.14  2007-03-16

- Added a few micro-optimizations from Ruslan Zakirov, who is hoping
  this will ultimately help speed up RT.


1.13  2006-04-01

- Add another fix for filename handling in the tests. Tests were
  giving false failures on Win32 because the tests needed to use
  File::Spec->canonpath(), just like Devel::StackTrace does
  internally.


1.12  2005-09-30

- Newer versions of Perl use Unix-style filenames when reporting the
  filename in caller(), which breaks Exception::Class tests on other
  platforms, and is just kind of funky.  This module now calls
  File::Spec->canonpath() to clean up the filename in each frame.

Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.023.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "An object representing a stack trace",
  "AUTHOR" => "Dave Rolsky <autarch\@urth.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Devel-StackTrace",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Devel::StackTrace",
  "PREREQ_PM" => {
    "File::Spec" => 0,
    "Scalar::Util" => 0,
    "overload" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.88",
    "base" => 0,
    "bytes" => 0
  },
  "VERSION" => "2.00",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Scalar::Util" => 0,
  "Test::More" => "0.88",
  "base" => 0,
  "bytes" => 0,
  "overload" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Devel/StackTrace/Frame.pm
package Devel::StackTrace::Frame;
$Devel::StackTrace::Frame::VERSION = '1.33';
use strict;
use warnings;

# Create accessor routines
BEGIN {
    no strict 'refs';
    foreach my $f (
        qw( package filename line subroutine hasargs
        wantarray evaltext is_require hints bitmask args )
        ) {
        next if $f eq 'args';
        *{$f} = sub { my $s = shift; return $s->{$f} };
    }
}

{
    my @fields = (
        qw( package filename line subroutine hasargs wantarray
            evaltext is_require hints bitmask )
    );

    sub new {
        my $proto = shift;
        my $class = ref $proto || $proto;

        my $self = bless {}, $class;

        @{$self}{@fields} = @{ shift() };

        # fixup unix-style paths on win32
        $self->{filename} = File::Spec->canonpath( $self->{filename} );

        $self->{args} = shift;

        $self->{respect_overload} = shift;

        $self->{max_arg_length} = shift;

        $self->{message} = shift;

        $self->{indent} = shift;

        return $self;
    }
}

sub args {
    my $self = shift;

    return @{ $self->{args} };
}

sub as_string {
    my $self  = shift;
    my $first = shift;
    my $p     = shift;

    my $sub = $self->subroutine;

    # This code stolen straight from Carp.pm and then tweaked.  All
    # errors are probably my fault  -dave
    if ($first) {
        $sub
            = defined $self->{message}
            ? $self->{message}
            : 'Trace begun';
    }
    else {

        # Build a string, $sub, which names the sub-routine called.
        # This may also be "require ...", "eval '...' or "eval {...}"
        if ( my $eval = $self->evaltext ) {
            if ( $self->is_require ) {
                $sub = "require $eval";
            }
            else {
                $eval =~ s/([\\\'])/\\$1/g;
                $sub = "eval '$eval'";
            }
        }
        elsif ( $sub eq '(eval)' ) {
            $sub = 'eval {...}';
        }

        # if there are any arguments in the sub-routine call, format
        # them according to the format variables defined earlier in
        # this file and join them onto the $sub sub-routine string
        #
        # We copy them because they're going to be modified.
        #
        if ( my @a = $self->args ) {
            for (@a) {

                # set args to the string "undef" if undefined
                $_ = "undef", next unless defined $_;

                # hack!
                $_ = $self->Devel::StackTrace::_ref_to_string($_)
                    if ref $_;

                local $SIG{__DIE__};
                local $@;

                eval {
                    my $max_arg_length
                        = exists $p->{max_arg_length}
                        ? $p->{max_arg_length}
                        : $self->{max_arg_length};

                    if ( $max_arg_length
                        && length $_ > $max_arg_length ) {
                        substr( $_, $max_arg_length ) = '...';
                    }

                    s/'/\\'/g;

                    # 'quote' arg unless it looks like a number
                    $_ = "'$_'" unless /^-?[\d.]+$/;

                    # print control/high ASCII chars as 'M-<char>' or '^<char>'
                    s/([\200-\377])/sprintf("M-%c",ord($1)&0177)/eg;
                    s/([\0-\37\177])/sprintf("^%c",ord($1)^64)/eg;
                };

                if ( my $e = $@ ) {
                    $_ = $e =~ /malformed utf-8/i ? '(bad utf-8)' : '?';
                }
            }

            # append ('all', 'the', 'arguments') to the $sub string
            $sub .= '(' . join( ', ', @a ) . ')';
            $sub .= ' called';
        }
    }

    # If the user opted into indentation (a la Carp::confess), pre-add a tab
    my $tab = $self->{indent} && !$first ? "\t" : q{};

    return "${tab}$sub at " . $self->filename . ' line ' . $self->line;
}

1;

# ABSTRACT: A single frame in a stack trace

__END__

=pod

=head1 NAME

Devel::StackTrace::Frame - A single frame in a stack trace

=head1 VERSION

version 1.33

=head1 DESCRIPTION

See L<Devel::StackTrace> for details.

=for Pod::Coverage new

=head1 METHODS

See Perl's C<caller()> documentation for more information on what these
methods return.

=head2 $frame->package()

=head2 $frame->filename()

=head2 $frame->line()

=head2 $frame->subroutine()

=head2 $frame->hasargs()

=head2 $frame->wantarray()

=head2 $frame->evaltext()

Returns undef if the frame was not part of an eval.

=head2 $frame->is_require()

Returns undef if the frame was not part of a require.

=head2 $frame->args()

Returns the arguments passed to the frame.  Note that any arguments that are
references are returned as references, not copies.

=head2 $frame->hints()

=head2 $frame->bitmask()

=head2 $frame->as_string()

Returns a string containing a description of the frame.

=head1 AUTHOR

Dave Rolsky <autarch@urth.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by Dave Rolsky.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


lib/Devel/StackTrace.pm
package Devel::StackTrace;
$Devel::StackTrace::VERSION = '1.33';
use 5.006;

use strict;
use warnings;

use Devel::StackTrace::Frame;
use File::Spec;
use Scalar::Util qw( blessed );

use overload
    '""'     => \&as_string,
    fallback => 1;

sub new {
    my $class = shift;
    my %p     = @_;

    # Backwards compatibility - this parameter was renamed to no_refs
    # ages ago.
    $p{no_refs} = delete $p{no_object_refs}
        if exists $p{no_object_refs};

    my $self = bless {
        index  => undef,
        frames => [],
        raw    => [],
        %p,
    }, $class;

    $self->_record_caller_data();

    return $self;
}

sub _record_caller_data {
    my $self = shift;

    my $filter = $self->{filter_frames_early} && $self->_make_frame_filter();

    # We exclude this method by starting at least one frame back.
    my $x = 1 + ( $self->{skip_frames} // 0 );

    while (
        my @c
        = $self->{no_args}
        ? caller( $x++ )
        : do {
            package    # the newline keeps dzil from adding a version here
                DB;
            @DB::args = ();
            caller( $x++ );
        }
        ) {

        my @args;

        @args = $self->{no_args} ? () : @DB::args;

        my $raw = {
            caller => \@c,
            args => \@args,
        };

        next if $filter && !$filter->($raw);

        if ( $self->{no_refs} ) {
            $raw->{args} = [ map { ref $_ ? $self->_ref_to_string($_) : $_ } @{$raw->{args}} ];
        }

        push @{ $self->{raw} }, $raw;
    }
}

sub _ref_to_string {
    my $self = shift;
    my $ref  = shift;

    return overload::AddrRef($ref)
        if blessed $ref && $ref->isa('Exception::Class::Base');

    return overload::AddrRef($ref) unless $self->{respect_overload};

    local $@;
    local $SIG{__DIE__};

    my $str = eval { $ref . '' };

    return $@ ? overload::AddrRef($ref) : $str;
}

sub _make_frames {
    my $self = shift;

    my $filter = !$self->{filter_frames_early} && $self->_make_frame_filter();

    my $raw = delete $self->{raw};
    for my $r ( @{$raw} ) {
        next if $filter && ! $filter->($r);

        $self->_add_frame( $r->{caller}, $r->{args} );
    }
}

my $default_filter = sub { 1 };

sub _make_frame_filter {
    my $self = shift;

    my ( @i_pack_re, %i_class );
    if ( $self->{ignore_package} ) {
        local $@;
        local $SIG{__DIE__};

        $self->{ignore_package} = [ $self->{ignore_package} ]
            unless eval { @{ $self->{ignore_package} } };

        @i_pack_re
            = map { ref $_ ? $_ : qr/^\Q$_\E$/ } @{ $self->{ignore_package} };
    }

    my $p = __PACKAGE__;
    push @i_pack_re, qr/^\Q$p\E$/;

    if ( $self->{ignore_class} ) {
        $self->{ignore_class} = [ $self->{ignore_class} ]
            unless ref $self->{ignore_class};
        %i_class = map { $_ => 1 } @{ $self->{ignore_class} };
    }

    my $user_filter = $self->{frame_filter};

    return sub {
        return 0 if grep { $_[0]{caller}[0] =~ /$_/ } @i_pack_re;
        return 0 if grep { $_[0]{caller}[0]->isa($_) } keys %i_class;

        if ($user_filter) {
            return $user_filter->( $_[0] );
        }

        return 1;
    };
}

sub _add_frame {
    my $self = shift;
    my $c    = shift;
    my $p    = shift;

    # eval and is_require are only returned when applicable under 5.00503.
    push @$c, ( undef, undef ) if scalar @$c == 6;

    push @{ $self->{frames} },
        Devel::StackTrace::Frame->new(
        $c,
        $p,
        $self->{respect_overload},
        $self->{max_arg_length},
        $self->{message},
        $self->{indent}
        );
}

sub next_frame {
    my $self = shift;

    # reset to top if necessary.
    $self->{index} = -1 unless defined $self->{index};

    my @f = $self->frames();
    if ( defined $f[ $self->{index} + 1 ] ) {
        return $f[ ++$self->{index} ];
    }
    else {
        $self->{index} = undef;
        return undef;
    }
}

sub prev_frame {
    my $self = shift;

    my @f = $self->frames();

    # reset to top if necessary.
    $self->{index} = scalar @f unless defined $self->{index};

    if ( defined $f[ $self->{index} - 1 ] && $self->{index} >= 1 ) {
        return $f[ --$self->{index} ];
    }
    else {
        $self->{index} = undef;
        return undef;
    }
}

sub reset_pointer {
    my $self = shift;

    $self->{index} = undef;
}

sub frames {
    my $self = shift;

    if (@_) {
        die
            "Devel::StackTrace->frames() can only take Devel::StackTrace::Frame args\n"
            if grep { !$_->isa('Devel::StackTrace::Frame') } @_;

        $self->{frames} = \@_;
    }
    else {
        $self->_make_frames() if $self->{raw};
    }

    return @{ $self->{frames} };
}

sub frame {
    my $self = shift;
    my $i    = shift;

    return unless defined $i;

    return ( $self->frames() )[$i];
}

sub frame_count {
    my $self = shift;

    return scalar( $self->frames() );
}

sub as_string {
    my $self = shift;
    my $p    = shift;

    my $st    = '';
    my $first = 1;
    foreach my $f ( $self->frames() ) {
        $st .= $f->as_string( $first, $p ) . "\n";
        $first = 0;
    }

    return $st;
}

{
    package
        Devel::StackTraceFrame;

    our @ISA = 'Devel::StackTrace::Frame';
}

1;

# ABSTRACT: An object representing a stack trace

__END__

=pod

=head1 NAME

Devel::StackTrace - An object representing a stack trace

=head1 VERSION

version 1.33

=head1 SYNOPSIS

  use Devel::StackTrace;

  my $trace = Devel::StackTrace->new();

  print $trace->as_string(); # like carp

  # from top (most recent) of stack to bottom.
  while ( my $frame = $trace->next_frame() ) {
      print "Has args\n" if $frame->hasargs();
  }

  # from bottom (least recent) of stack to top.
  while ( my $frame = $trace->prev_frame() ) {
      print "Sub: ", $frame->subroutine(), "\n";
  }

=head1 DESCRIPTION

The C<Devel::StackTrace> module contains two classes, C,Devel::StackTrace> and
L<Devel::StackTrace::Frame>. These objects encapsulate the information that
can retrieved via Perl's C<caller()> function, as well as providing a simple
interface to this data.

The C<Devel::StackTrace> object contains a set of C<Devel::StackTrace::Frame>
objects, one for each level of the stack. The frames contain all the data
available from C<caller()>.

This code was created to support my L<Exception::Class::Base> class (part of
L<Exception::Class>) but may be useful in other contexts.

=head1 'TOP' AND 'BOTTOM' OF THE STACK

When describing the methods of the trace object, I use the words 'top' and
'bottom'. In this context, the 'top' frame on the stack is the most recent
frame and the 'bottom' is the least recent.

Here's an example:

  foo();  # bottom frame is here

  sub foo {
     bar();
  }

  sub bar {
     Devel::StackTrace->new();  # top frame is here.
  }

=head1 METHODS

This class provide the following methods:

=head2 Devel::StackTrace->new(%named_params)

Returns a new Devel::StackTrace object.

Takes the following parameters:

=over 4

=item * frame_filter => $sub

By default, Devel::StackTrace will include all stack frames before the
call to its constructor.

However, you may want to filter out some frames with more granularity
than 'ignore_package' or 'ignore_class' allow.

You can provide a subroutine which is called with the raw frame data
for each frame. This is a hash reference with two keys, "caller", and
"args", both of which are array references. The "caller" key is the
raw data as returned by Perl's C<caller()> function, and the "args"
key are the subroutine arguments found in C<@DB::args>.

The filter should return true if the frame should be included, or
false if it should be skipped.

=item * filter_frames_early => $boolean

If this parameter is true, C<frame_filter> will be called as soon as the
stacktrace is created, and before refs are stringified (if C<no_refs> is
true), rather than being filtered lazily when L<Devel::StackTrace::Frame>
objects are first needed.

This is useful if you want to filter based on the frame's arguments and want
to be able to examine object properties, for example.

=item * ignore_package => $package_name OR \@package_names

Any frames where the package is one of these packages will not be on
the stack.

=item * ignore_class => $package_name OR \@package_names

Any frames where the package is a subclass of one of these packages
(or is the same package) will not be on the stack.

Devel::StackTrace internally adds itself to the 'ignore_package'
parameter, meaning that the Devel::StackTrace package is B<ALWAYS>
ignored. However, if you create a subclass of Devel::StackTrace it
will not be ignored.

=item * skip_frames => $integer

This will cause this number of stack frames to be excluded from top of the
stack trace. This prevents the frames from being captured at all, and applies
before the C<frame_filter>, C<ignore_package>, or C<ignore_class> options,
even with C<filter_frames_early>.

=item * no_refs => $boolean

If this parameter is true, then Devel::StackTrace will not store
references internally when generating stacktrace frames. This lets
your objects go out of scope.

Devel::StackTrace replaces any references with their stringified
representation.

=item * no_args => $boolean

If this parameter is true, then Devel::StackTrace will not store caller
arguments in stack trace frames at all.

=item * respect_overload => $boolean

By default, Devel::StackTrace will call C<overload::AddrRef()> to get
the underlying string representation of an object, instead of
respecting the object's stringification overloading. If you would
prefer to see the overloaded representation of objects in stack
traces, then set this parameter to true.

=item * max_arg_length => $integer

By default, Devel::StackTrace will display the entire argument for each
subroutine call. Setting this parameter causes truncates each subroutine
argument's string representation if it is longer than this number of
characters.

=item * message => $string

By default, Devel::StackTrace will use 'Trace begun' as the message for the
first stack frame when you call C<as_string>. You can supply an alternative
message using this option.

=item * indent => $boolean

If this parameter is true, each stack frame after the first will start with a
tab character, just like C<Carp::confess()>.

=back

=head2 $trace->next_frame()

Returns the next L<Devel::StackTrace::Frame> object on the stack, going
down. If this method hasn't been called before it returns the first frame. It
returns C<undef> when it reaches the bottom of the stack and then resets its
pointer so the next call to C<< $trace->next_frame() >> or C<<
$trace->prev_frame() >> will work properly.

=head2 $trace->prev_frame()

Returns the next L<Devel::StackTrace::Frame> object on the stack, going up. If
this method hasn't been called before it returns the last frame. It returns
undef when it reaches the top of the stack and then resets its pointer so the
next call to C<< $trace->next_frame() >> or C<< $trace->prev_frame() >> will
work properly.

=head2 $trace->reset_pointer

Resets the pointer so that the next call to C<< $trace->next_frame() >> or C<<
$trace->prev_frame() >> will start at the top or bottom of the stack, as
appropriate.

=head2 $trace->frames()

When this method is called with no arguments, it returns a list of
L<Devel::StackTrace::Frame> objects. They are returned in order from top (most
recent) to bottom.

This method can also be used to set the object's frames if you pass it a list
of L<Devel::StackTrace::Frame> objects.

This is useful if you want to filter the list of frames in ways that are more
complex than can be handled by the C<< $trace->filter_frames() >> method:

  $stacktrace->frames( my_filter( $stacktrace->frames() ) );

=head2 $trace->frame($index)

Given an index, this method returns the relevant frame, or undef if there is
no frame at that index. The index is exactly like a Perl array. The first
frame is 0 and negative indexes are allowed.

=head2 $trace->frame_count()

Returns the number of frames in the trace object.

=head2 $trace->as_string(\%p)

Calls C<< $frame->as_string() >> on each frame from top to bottom, producing
output quite similar to the Carp module's cluck/confess methods.

The optional C<\%p> parameter only has one option. The C<max_arg_length>
parameter truncates each subroutine argument's string representation if it is
longer than this number of characters.

=head1 SUPPORT

Please submit bugs to the CPAN RT system at
http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Devel%3A%3AStackTrace
or via email at bug-devel-stacktrace@rt.cpan.org.

=head1 AUTHOR

Dave Rolsky <autarch@urth.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by Dave Rolsky.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut
